### ISSUE TITLE
N CIGAR code to be treated as D

### ISSUE THREAD

**Issue by iosonofabio** [freebayes/freebayes#729]
**Describe the bug**

Thank you for this amazing package! Running `freebayes` on a typical RNA-Seq data (BAM file) set fails when a read has an `N` CIGAR code, which means "skip". "skip" is technically equivalent to `D` (deletion), they are kept separate in the SAM file spec because their biological interpretation is different.

However, as far as calling SNPs is concerned, the user does not care about either code, and having `freebayes` dying on you makes it necessary to convert `N` into `D` throughout the BAM file which is time consuming and creates additional complexity.

**To Reproduce**

Happy to provide an example if you think this is worth looking into. Any BAM file with reads that include the `N` CIGAR code will reproduce this. Typical call:

```bash
freebayes-parallel /home/fabio/projects/dengue_patients/sequencing_data/vcf/genome_regions_freebayes.fa.fai 32 -f /home/fabio/software/cellranger_reference_data/refdata-gex-GRCh38-2020-A/fasta/genome.fa -iXu -C 2 -q 1 /home/fabio/projects/dengue_patients/sequencing_data/vcf/20211218_4adults_duplexed/H_3008_3024/possorted_genome_bam_NtoD.bam > /home/fabio/projects/dengue_patients/sequencing_data/vcf/20211218_4adults_duplexed/H_3008_3024/freebayes_snv.vcf
```

(same happens with the non-parallel version).

**Expected behavior**

Either an option to treat `N` and `D`, or silent equivalence.

**Finally**

Happy to make a PR if you think this is useful.

Thanks

**Comment by iosonofabio** [comment]
Honored to try and give up halfway through 🙂 give me a few days

**Comment by iosonofabio** [comment]
I started this, but things look dire and it would be fantastic if you could help a little.

I peeked into your `.github` workflows. Seems like one has to update all git submodules, so I did. Then I ran `meson` to configure and `ninja` to build.

Configure complains about the absence of a bunch of libraries:

```bash
The Meson build system
Version: 0.61.1
Source dir: /home/fabio/university/PI/projects/freebayes
Build dir: /home/fabio/university/PI/projects/freebayes/build
Build type: native build
Project name: freebayes
Project version: 1.3.6
C compiler for the host machine: cc (gcc 11.1.0 "cc (GCC) 11.1.0")
C linker for the host machine: cc ld.bfd 2.36.1
C++ compiler for the host machine: c++ (gcc 11.1.0 "c++ (GCC) 11.1.0")
C++ linker for the host machine: c++ ld.bfd 2.36.1
Host machine cpu family: x86_64
Host machine cpu: x86_64
Dependency zlib found: YES 1.2.11 (cached)
Dependency liblzma found: YES 5.2.5 (cached)
Dependency threads found: YES unknown (cached)
Library tabixpp found: NO
Dependency htslib found: YES 1.14 (cached)
Library tabixpp found: NO
Found pkg-config: /usr/bin/pkg-config (1.8.0)
Found CMake: /usr/bin/cmake (3.22.1)
Run-time dependency libvcflib found: NO (tried pkgconfig and cmake)
Run-time dependency libseqlib found: NO (tried pkgconfig and cmake)
Compiler for C++ supports arguments -fpermissive: YES (cached)
Compiler for C++ supports arguments -Wno-reorder: YES (cached)
Compiler for C++ supports arguments -Wno-sign-compare: YES (cached)
Compiler for C++ supports arguments -Wno-unused-variable: YES (cached)
Compiler for C++ supports arguments -Wno-unused-but-set-variable: YES (cached)
Program prove found: YES (/usr/bin/core_perl/prove)
Build targets in project: 6

Found ninja-1.10.2 at /usr/bin/ninja
```
Meson seems confused about when to use system shared libraries versus the submodules. For instance, it cannot find `tabixpp` or `vcflib`, even though it made a folder called `vcflib` through the submodule command, and that folder contains a subfolder called `tabixpp` which itself contains a bunch of files including the supposedly missing header `tabix.hpp`.

Of course, I tried to go ahead anyway, ninja fails with some unexpected complaints, e.g.:

```bash
In file included from ../src/BedReader.h:12,
                 from ../src/AlleleParser.h:20,
                 from ../src/AlleleParser.cpp:1:
../src/../intervaltree/IntervalTree.h:274:75: error: ‘numeric_limits’ is not a member of ‘std’
  274 |                                                                      std::numeric_limits<Scalar>::min() }};
      |                                                                           ^~~~~~~~~~~~~~
../src/../intervaltree/IntervalTree.h:274:96: error: expected primary-expression before ‘>’ token
  274 |                                                                      std::numeric_limits<Scalar>::min() }};
      |                                                                                                ^
../src/../intervaltree/IntervalTree.h:274:99: error: ‘::min’ has not been declared; did you mean ‘std::min’?
  274 |                                                                      std::numeric_limits<Scalar>::min() }};
      |                                                                                                   ^~~
      |                                                                                                   std::min
```

Now, intervaltree is a submodule, so the fact that its header fails on C++ standard library functions makes it a little difficult to debug.

Any suggestions?

**Comment by pjotrp** [comment]
To use the local source code checkouts use 

    meson build -Dprefer_system_deps=false

after removing the build directory.

If you still get the C++ errors I suspect it is coming from GNU gcc
11. I have been runnning version 10. Let me see if I get the same on
that one with v11.

**Comment by iosonofabio** [comment]
Thank you! Still can't find `tabixpp`:

```bash
[freebayes] (master)$ meson build -Dprefer_system_deps=false
The Meson build system
Version: 0.61.1
Source dir: /home/fabio/university/PI/projects/freebayes
Build dir: /home/fabio/university/PI/projects/freebayes/build
Build type: native build
Project name: freebayes
Project version: 1.3.6
C compiler for the host machine: cc (gcc 11.1.0 "cc (GCC) 11.1.0")
C linker for the host machine: cc ld.bfd 2.36.1
C++ compiler for the host machine: c++ (gcc 11.1.0 "c++ (GCC) 11.1.0")
C++ linker for the host machine: c++ ld.bfd 2.36.1
Host machine cpu family: x86_64
Host machine cpu: x86_64
Found pkg-config: /usr/bin/pkg-config (1.8.0)
Run-time dependency zlib found: YES 1.2.11
Run-time dependency liblzma found: YES 5.2.5
Run-time dependency threads found: YES
Library tabixpp found: NO
Compiler for C++ supports arguments -fpermissive: YES 
Compiler for C++ supports arguments -Wno-reorder: YES 
Compiler for C++ supports arguments -Wno-sign-compare: YES 
Compiler for C++ supports arguments -Wno-unused-variable: YES 
Compiler for C++ supports arguments -Wno-unused-but-set-variable: YES 
Program prove found: YES (/usr/bin/core_perl/prove)
Build targets in project: 7

freebayes 1.3.6

  User defined options
    prefer_system_deps: false

Found ninja-1.10.2 at /usr/bin/ninja
```

and if I ignore and run ninja:

```bash
[freebayes] (master)$ ninja -C build                                                                
ninja: Entering directory `build'
[19/81] Compiling C object libcustom_htslib.a.p/contrib_htslib_htscodecs_htscodecs_htscodecs.c.o
FAILED: libcustom_htslib.a.p/contrib_htslib_htscodecs_htscodecs_htscodecs.c.o 
cc -Ilibcustom_htslib.a.p -I. -I.. -I../contrib/htslib -I../contrib/htslib_config -I../contrib/htslib/cram -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -O3 -g -fPIC -MD -MQ libcustom_htslib.a.p/contrib_htslib_htscodecs_htscodecs_htscodecs.c.o -MF libcustom_htslib.a.p/contrib_htslib_htscodecs_htscodecs_htscodecs.c.o.d -o libcustom_htslib.a.p/contrib_htslib_htscodecs_htscodecs_htscodecs.c.o -c ../contrib/htslib/htscodecs/htscodecs/htscodecs.c
../contrib/htslib/htscodecs/htscodecs/htscodecs.c: In function ‘htscodecs_version’:
../contrib/htslib/htscodecs/htscodecs/htscodecs.c:43:12: error: ‘HTSCODECS_VERSION_TEXT’ undeclared (first use in this function); did you mean ‘HTSCODECS_VERSION’?
   43 |     return HTSCODECS_VERSION_TEXT;
      |            ^~~~~~~~~~~~~~~~~~~~~~
      |            HTSCODECS_VERSION
../contrib/htslib/htscodecs/htscodecs/htscodecs.c:43:12: note: each undeclared identifier is reported only once for each function it appears in
[28/81] Compiling C object libcustom_htslib.a.p/contrib_htslib_sam.c.o
ninja: build stopped: subcommand failed.
```

I guess having a different error is progress ;-)

**Comment by pjotrp** [comment]
tabixpp is part of vcflib checkout. You need to init the modules
recursively.

    git submodule update --init --recursive --progress

See the README.

Alternatively install it as a distro package.

**Comment by pjotrp** [comment]
See also for the error you should have https://github.com/freebayes/freebayes/blob/master/contrib/htslib_config/version.h

**Comment by pjotrp** [comment]
I am getting a build error with gcc 11. Looking into that.

**Comment by pjotrp** [comment]
@iosonofabio I actually use the Guix packages to write code. It gives full control over dependencies. If you want to try simply install guix (it is part of Debian), run a `guix pull` and see [guix.scm](./guix.scm) for running a container.

**Comment by iosonofabio** [comment]
Thanks @pjotrp I pulled your commit, updated the submodules recursively, and now can compile the master branch.

FWIW, `meson` still says I have no `tabixpp`:

```bash
[freebayes] (master)$ meson build --reconfigure -Dprefer_system_deps=false
The Meson build system
Version: 0.61.1
Source dir: /home/fabio/university/PI/projects/freebayes
Build dir: /home/fabio/university/PI/projects/freebayes/build
Build type: native build
Project name: freebayes
Project version: 1.3.6
C compiler for the host machine: cc (gcc 11.1.0 "cc (GCC) 11.1.0")
C linker for the host machine: cc ld.bfd 2.36.1
C++ compiler for the host machine: c++ (gcc 11.1.0 "c++ (GCC) 11.1.0")
C++ linker for the host machine: c++ ld.bfd 2.36.1
Host machine cpu family: x86_64
Host machine cpu: x86_64
Dependency zlib found: YES 1.2.11 (cached)
Dependency liblzma found: YES 5.2.5 (cached)
Dependency threads found: YES unknown (cached)
Library tabixpp found: NO
Compiler for C++ supports arguments -fpermissive: YES (cached)
Compiler for C++ supports arguments -Wno-reorder: YES (cached)
Compiler for C++ supports arguments -Wno-sign-compare: YES (cached)
Compiler for C++ supports arguments -Wno-unused-variable: YES (cached)
Compiler for C++ supports arguments -Wno-unused-but-set-variable: YES (cached)
Program prove found: YES (/usr/bin/core_perl/prove)
Build targets in project: 7

freebayes 1.3.6

  User defined options
    prefer_system_deps: false

Found ninja-1.10.2 at /usr/bin/ninja
```

but then `ninja -C build` works fine, so that looks like an inconsistency/bug to me.

I'm not testing the compiled binary on a few files that are made by `STAR` and have `N` (null allele) at introns and otherwise choke `freebayes`. Will keep you posted.

I looked into `guix`, looks like the vegan new kid around the block. I'm not terribly thrilled to mess with package managers beyond `pacman`, `conda`, `pip`, `npm`, which are bad enough, so I'll take a rain check for now unless absolutely necessary. But thanks for the suggestion, of course I might give it a go when I'm not in a total rush.

**Comment by pjotrp** [comment]
Tabixpp comes with the vcflib module. meson just says it can't find it on your system (it would use the system version by default).

GNU Guix is not a normal package manager btw. For one it installs files only in /etc/guix and /gnu/store. Remove those dirs and it is gone! No interference with anything. Guaranteed.

**Comment by iosonofabio** [comment]
Thanks. I now discovered I was using an older version of `freebayes`. It seems this bug was already solved somewhere between 0.9.1 and 1.3.5, since the newest version does not choke on introns. It gives a weird message about not having enough file handles, but that's probably unrelated.

Therefore, I'll close this for now, thanks for everything! If I'm wrong and this was NOT fixed, I'll reopen ;-)

**Comment by pjotrp** [comment]
Ah, this is somewhat deflating :). Always use the latest when reporting an issue. @iosonofabio can you reproduce the file handle problem? Maybe open a new bug if it is not on the issue tracker. Running out of file handles is an OS-type issue - usually.
